File: routes/patient.py
from flask import Blueprint, request, jsonify, session
from datetime import datetime
from database import (
    get_patient_by_email, get_patient_by_id, create_patient, update_patient,
    get_all_appointments, get_appointments_by_patient,
    get_appointment_by_id, create_appointment, update_appointment
)

patient_bp = Blueprint("patient", __name__)


def _require_patient():
    user = session.get("patient")
    if not user:
        return None
    return user


# ── Patient auth ───────────────────────────────────────

@patient_bp.route("/api/patient/register", methods=["POST"])
def patient_register():
    d = request.get_json()
    required = ["fname","lname","mname","age","sex","birthday",
                "address","phone","email","messenger","password"]
    missing = [f for f in required if not d.get(f)]
    if missing:
        return jsonify({"error": f"Please fill in all required fields: {', '.join(missing)}"}), 400

    email = d["email"].strip().lower()
    if get_patient_by_email(email):
        return jsonify({"error": "Email is already registered. Please log in."}), 400

    # Basic validations
    import re
    if not re.match(r"[^@]+@[^@]+\.[^@]+", email):
        return jsonify({"error": "Invalid email format."}), 400
    if not str(d["age"]).isdigit() or int(d["age"]) < 1 or int(d["age"]) > 120:
        return jsonify({"error": "Age must be a valid number."}), 400

    patient = create_patient({
        "fname":       d["fname"].strip(),
        "lname":       d["lname"].strip(),
        "mname":       d["mname"].strip(),
        "name":        f"{d['fname'].strip()} {d['mname'].strip()} {d['lname'].strip()}",
        "age":         int(d["age"]),
        "sex":         d["sex"],
        "birthday":    d["birthday"],
        "address":     d["address"].strip(),
        "phone":       d["phone"].strip(),
        "email":       email,
        "messenger":   d["messenger"].strip(),
        "password":    d["password"],
        "registered_at": datetime.now().strftime("%Y-%m-%d %H:%M"),
    })
    session["patient"] = {"id": patient["id"], "name": patient["name"],
                          "email": patient["email"], "role": "patient"}
    return jsonify({"ok": True, "name": patient["name"]})


@patient_bp.route("/api/patient/login", methods=["POST"])
def patient_login():
    d = request.get_json()
    email    = d.get("email","").strip().lower()
    password = d.get("password","")
    patient  = get_patient_by_email(email)
    if not patient or patient["password"] != password:
        return jsonify({"error": "Invalid email or password."}), 401
    session["patient"] = {"id": patient["id"], "name": patient["name"],
                          "email": patient["email"], "role": "patient"}
    return jsonify({"ok": True, "name": patient["name"]})


@patient_bp.route("/api/patient/logout", methods=["POST"])
def patient_logout():
    session.pop("patient", None)
    return jsonify({"ok": True})


@patient_bp.route("/api/patient/me")
def patient_me():
    p = _require_patient()
    if not p:
        return jsonify({"error": "Not logged in"}), 401
    patient = get_patient_by_id(p["id"])
    return jsonify(patient)


# ── Appointments ───────────────────────────────────────

@patient_bp.route("/api/patient/appointments", methods=["GET"])
def get_my_appointments():
    p = _require_patient()
    if not p:
        return jsonify({"error": "Unauthorized"}), 401
    return jsonify(get_appointments_by_patient(p["id"]))


@patient_bp.route("/api/patient/appointments", methods=["POST"])
def book_appointment():
    p = _require_patient()
    if not p:
        return jsonify({"error": "Unauthorized"}), 401

    d = request.get_json()
    required = ["appt_date", "appt_time", "concern"]
    missing = [f for f in required if not d.get(f)]
    if missing:
        return jsonify({"error": f"Missing: {', '.join(missing)}"}), 400

    patient = get_patient_by_id(p["id"])
    appt = create_appointment({
        "patient_id":   p["id"],
        "patient_name": patient["name"],
        "patient_email":patient["email"],
        "patient_phone":patient.get("phone",""),
        "patient_messenger": patient.get("messenger",""),
        "appt_date":    d["appt_date"],
        "appt_time":    d["appt_time"],
        "concern":      d["concern"].strip(),
        "status":       "Pending",
        "urgent":       d.get("urgent", False),
        "notes":        "",
        "submitted_at": datetime.now().strftime("%Y-%m-%d %H:%M"),
        "updated_at":   datetime.now().strftime("%Y-%m-%d %H:%M"),
    })
    return jsonify({"ok": True, "id": appt["id"]})


@patient_bp.route("/api/patient/appointments/<int:aid>", methods=["PUT"])
def update_my_appointment(aid):
    p = _require_patient()
    if not p:
        return jsonify({"error": "Unauthorized"}), 401
    appt = get_appointment_by_id(aid)
    if not appt or appt["patient_id"] != p["id"]:
        return jsonify({"error": "Not found"}), 404
    if appt["status"] == "Completed":
        return jsonify({"error": "Cannot edit a completed appointment."}), 400

    d = request.get_json()
    updates = {}
    for field in ["appt_date","appt_time","concern"]:
        if d.get(field):
            updates[field] = d[field]
    updates["updated_at"] = datetime.now().strftime("%Y-%m-%d %H:%M")
    update_appointment(aid, updates)
    return jsonify({"ok": True})


@patient_bp.route("/api/patient/appointments/<int:aid>/cancel", methods=["POST"])
def cancel_appointment(aid):
    p = _require_patient()
    if not p:
        return jsonify({"error": "Unauthorized"}), 401
    appt = get_appointment_by_id(aid)
    if not appt or appt["patient_id"] != p["id"]:
        return jsonify({"error": "Not found"}), 404
    update_appointment(aid, {"status": "Cancelled",
                              "updated_at": datetime.now().strftime("%Y-%m-%d %H:%M")})
    return jsonify({"ok": True})


# ── Admin: manage appointments ─────────────────────────

@patient_bp.route("/api/admin/appointments", methods=["GET"])
def admin_get_appointments():
    from flask import session as s
    user = s.get("user")
    if not user or user.get("role") not in ("admin","nurse"):
        return jsonify({"error": "Forbidden"}), 403
    appts = get_all_appointments()
    # sort: Pending first, then by date
    appts.sort(key=lambda a: (
        0 if a["status"]=="Pending" else 1 if a["status"]=="Confirmed" else 2,
        a.get("appt_date","")
    ))
    return jsonify(appts)


@patient_bp.route("/api/admin/appointments/<int:aid>/status", methods=["POST"])
def admin_update_status(aid):
    from flask import session as s
    user = s.get("user")
    if not user or user.get("role") not in ("admin","nurse"):
        return jsonify({"error": "Forbidden"}), 403
    d = request.get_json()
    status = d.get("status")
    if status not in ("Pending","Confirmed","Completed","Cancelled"):
        return jsonify({"error": "Invalid status"}), 400
    updates = {"status": status,
               "updated_at": datetime.now().strftime("%Y-%m-%d %H:%M")}
    if d.get("notes"):
        updates["notes"] = d["notes"]
    update_appointment(aid, updates)
    return jsonify({"ok": True})


@patient_bp.route("/api/admin/patients", methods=["GET"])
def admin_get_patients():
    from flask import session as s
    from database import get_all_patients
    user = s.get("user")
    if not user or user.get("role") not in ("admin","nurse"):
        return jsonify({"error": "Forbidden"}), 403
    patients = get_all_patients()
    # remove passwords
    return jsonify([{k:v for k,v in p.items() if k != "password"} for p in patients])

File: static/js/patient.js
/* ═══════════════════════════════════════════════════════
   patient.js — Patient portal: register, login, appointments
   ═══════════════════════════════════════════════════════ */

let _patientUser = null;

/* ── Auth ─────────────────────────────────────────────── */
async function patientRegister() {
  const fields = {
    fname:     val("preg-fname"),    lname:    val("preg-lname"),
    mname:     val("preg-mname"),    age:      val("preg-age"),
    sex:       val("preg-sex"),      birthday: val("preg-birthday"),
    address:   val("preg-address"),  phone:    val("preg-phone"),
    email:     val("preg-email"),    messenger:val("preg-messenger"),
    password:  val("preg-password"), confirm:  val("preg-confirm"),
    appt_date: val("preg-appt-date"),appt_time:val("preg-appt-time"),
    concern:   val("preg-concern"),
  };

  // Validate all required
  const reqFields = ["fname","lname","mname","age","sex","birthday",
                     "address","phone","email","messenger","password","confirm",
                     "appt_date","appt_time","concern"];
  for (const f of reqFields) {
    if (!fields[f]) { toast(`Please fill in all required fields.`, "error"); return; }
  }
  if (fields.password !== fields.confirm) {
    toast("Passwords do not match.", "error"); return;
  }
  if (!/^[^\s@]+@[^\s@]+\.[^\s@]+$/.test(fields.email)) {
    toast("Please enter a valid email address.", "error"); return;
  }
  if (isNaN(fields.age) || +fields.age < 1 || +fields.age > 120) {
    toast("Please enter a valid age.", "error"); return;
  }

  const res = await api("/api/patient/register", "POST", fields);
  if (res.error) { toast(res.error, "error"); return; }

  // Auto-book appointment after register
  await api("/api/patient/appointments", "POST", {
    appt_date: fields.appt_date,
    appt_time: fields.appt_time,
    concern:   fields.concern,
  });

  _patientUser = { name: res.name };
  toast(`Welcome, ${res.name}! Your appointment request has been submitted.`, "success");
  await loadPatientPortal();
  showPage("patient-portal");
}

async function patientLogin() {
  const email    = val("plogin-email");
  const password = val("plogin-password");
  if (!email || !password) { toast("Please enter email and password.", "error"); return; }

  const res = await api("/api/patient/login", "POST", { email, password });
  if (res.error) { toast(res.error, "error"); return; }

  _patientUser = { name: res.name };
  toast(`Welcome back, ${res.name}!`, "success");
  await loadPatientPortal();
  showPage("patient-portal");
}

async function patientLogout() {
  await api("/api/patient/logout", "POST");
  _patientUser = null;
  showPage("home");
  updateNav();
}

/* ── Portal ───────────────────────────────────────────── */
async function loadPatientPortal() {
  if (!_patientUser) return;

  document.getElementById("portal-name").textContent = _patientUser.name;

  const me   = await api("/api/patient/me");
  const appts = await api("/api/patient/appointments");

  // Fill profile display
  const fields = ["fname","lname","mname","age","sex","birthday","address","phone","email","messenger"];
  fields.forEach(f => {
    const el = document.getElementById("prof-" + f);
    if (el) el.textContent = me[f] || "—";
  });

  renderPatientAppointments(appts);
}

function renderPatientAppointments(appts) {
  const tbody = document.getElementById("patient-appt-tbody");
  if (!tbody) return;

  if (!appts.length) {
    tbody.innerHTML = `<tr><td colspan="6" style="text-align:center;color:var(--muted);padding:2rem">No appointments yet.</td></tr>`;
    return;
  }

  tbody.innerHTML = [...appts].reverse().map(a => `
    <tr>
      <td><strong>${a.appt_date}</strong></td>
      <td>${a.appt_time}</td>
      <td style="max-width:200px;font-size:.83rem">${a.concern}</td>
      <td>${a.submitted_at}</td>
      <td><span class="appt-badge appt-${a.status.toLowerCase()}">${a.status}</span>
          ${a.notes ? `<div style="font-size:.72rem;color:var(--muted);margin-top:3px">Note: ${a.notes}</div>` : ""}
      </td>
      <td>
        ${a.status === "Pending" ? `
          <button class="action-btn btn-view" onclick="openEditAppt(${a.id},'${a.appt_date}','${a.appt_time}',\`${a.concern.replace(/`/g,"'")}\`)">✏️ Edit</button>
          <button class="action-btn btn-reject" onclick="cancelAppt(${a.id})" style="margin-left:4px">✕ Cancel</button>
        ` : `<span style="color:var(--muted);font-size:.78rem">${a.status}</span>`}
      </td>
    </tr>`).join("");
}

function openBookAppt() {
  openModal(`
    <h2 class="modal-title">📅 Book New Appointment</h2>
    <div class="form-group"><label class="form-label">Appointment Date *</label>
      <input class="form-input" type="date" id="new-appt-date" min="${new Date().toISOString().split('T')[0]}"></div>
    <div class="form-group"><label class="form-label">Appointment Time *</label>
      <input class="form-input" type="time" id="new-appt-time"></div>
    <div class="form-group"><label class="form-label">Concern / Message *</label>
      <textarea class="form-textarea" id="new-appt-concern" placeholder="Describe your concern..." style="min-height:100px"></textarea></div>
    <div class="form-group"><label class="chk-label"><input type="checkbox" id="new-appt-urgent"> 🔴 Mark as urgent</label></div>
    <div class="submit-row">
      <button class="btn-cancel" onclick="closeModal()">Cancel</button>
      <button class="btn-primary" onclick="submitNewAppt()">Submit Request</button>
    </div>`);
}

async function submitNewAppt() {
  const appt_date = val("new-appt-date");
  const appt_time = val("new-appt-time");
  const concern   = val("new-appt-concern");
  const urgent    = document.getElementById("new-appt-urgent")?.checked || false;

  if (!appt_date || !appt_time || !concern) {
    toast("Please fill in all fields.", "error"); return;
  }
  const res = await api("/api/patient/appointments", "POST", {appt_date, appt_time, concern, urgent});
  if (res.error) { toast(res.error, "error"); return; }
  toast("Appointment request submitted!", "success");
  closeModal();
  const appts = await api("/api/patient/appointments");
  renderPatientAppointments(appts);
}

function openEditAppt(id, date, time, concern) {
  openModal(`
    <h2 class="modal-title">✏️ Edit Appointment</h2>
    <div class="form-group"><label class="form-label">Appointment Date *</label>
      <input class="form-input" type="date" id="edit-appt-date" value="${date}" min="${new Date().toISOString().split('T')[0]}"></div>
    <div class="form-group"><label class="form-label">Appointment Time *</label>
      <input class="form-input" type="time" id="edit-appt-time" value="${time}"></div>
    <div class="form-group"><label class="form-label">Concern / Message *</label>
      <textarea class="form-textarea" id="edit-appt-concern" style="min-height:100px">${concern}</textarea></div>
    <div class="submit-row">
      <button class="btn-cancel" onclick="closeModal()">Cancel</button>
      <button class="btn-primary" onclick="saveEditAppt(${id})">Save Changes</button>
    </div>`);
}

async function saveEditAppt(id) {
  const appt_date = val("edit-appt-date");
  const appt_time = val("edit-appt-time");
  const concern   = val("edit-appt-concern");
  if (!appt_date || !appt_time || !concern) {
    toast("Please fill in all fields.", "error"); return;
  }
  const res = await api(`/api/patient/appointments/${id}`, "PUT", {appt_date, appt_time, concern});
  if (res.error) { toast(res.error, "error"); return; }
  toast("Appointment updated!", "success");
  closeModal();
  const appts = await api("/api/patient/appointments");
  renderPatientAppointments(appts);
}

async function cancelAppt(id) {
  if (!confirm("Cancel this appointment request?")) return;
  const res = await api(`/api/patient/appointments/${id}/cancel`, "POST");
  if (res.error) { toast(res.error, "error"); return; }
  toast("Appointment cancelled.", "info");
  const appts = await api("/api/patient/appointments");
  renderPatientAppointments(appts);
}
